redis_collections/dicts.py

```diff
@@ -122,70 +120,61 @@ def __getitem__(self, key):
-            items is not very efficient. If possible, use :func:`get`.
         """
-        with self.redis.pipeline() as pipe:
-            pipe.hexists(self.key, key)
-            pipe.hget(self.key, key)
-            exists, value = pipe.execute()
+        key_hash, D = self._get_hash_dict(key, self.redis)
 
-        if not exists:
+        try:
+            value = D[key]
+        except KeyError:
             if hasattr(self, '__missing__'):
                 return self.__missing__(key)
-            raise KeyError(key)
-        return self._unpickle(value)
+            else:
+                raise
+
+        return value
 
     def __setitem__(self, key, value):
         """Set ``d[key]`` to *value*."""
-        value = self._pickle(value)
-        self.redis.hset(self.key, key, value)
+        key_hash, D = self._get_hash_dict(key, self.redis)
+        D[key] = value
 
-    def __delitem__(self, key):
-        """Remove ``d[key]`` from dictionary. Raises
-        a :func:`KeyError` if *key* is not in the map.
+        self.redis.hset(self.key, key_hash, self._pickle(D))
 
-        .. note::
-            Due to implementation on Redis side, this method of deleting
-            items is not very efficient. If possible, use :func:`discard`.
+    def __delitem__(self, key):
+        """Remove ``d[key]`` from dictionary.
+        Raises a :func:`KeyError` if *key* is not in the map.
         """
-        with self.redis.pipeline() as pipe:
-            pipe.hexists(self.key, key)
-            pipe.hdel(self.key, key)
-            exists, _ = pipe.execute()
+        key_hash, D = self._get_hash_dict(key, self.redis)
+        del D[key]
 
-        if not exists:
-            raise KeyError(key)
+        if D:
+            self.redis.hset(self.key, key_hash, self._pickle(D))
+        else:
+            self.redis.hdel(self.key, key_hash)
 
-    def discard(self, key):
-        """Remove ``d[key]`` from dictionary if it is present.
+    def _data(self, pipe=None):
+        """Returns a Python dictionary with the same values as this object"""
+        redis = self.redis if pipe is None else pipe
 
-        .. note::
-            Due to implementation on Redis side, this method of retrieving
-            items is more efficient than classic approach over using the
-            :func:`__delitem__` protocol.
-        """
-        self.redis.hdel(self.key, key)
+        ret = {}
+        for D in six.itervalues(redis.hgetall(self.key)):
+            ret.update(self._unpickle(D))
 
-    def _data(self, pipe=None):
-        redis = pipe if pipe is not None else self.redis
-        result = redis.hgetall(self.key).items()
-        return [(k.decode('utf-8'), self._unpickle(v)) for (k, v) in result]
+        return ret
 
     def items(self):
         """Return a copy of the dictionary's list of ``(key, value)`` pairs."""
-        return self._data()
+        return list(self.iteritems())
```

Comment:
In the future I may implement [dictview](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects) for these collections... but not now.